### PR TITLE
fix: Fix mypy type checking

### PR DESCRIPTION
**Pull request by robinholzi**
Previously some `__init__.py` were missing. Therefore mypy didn't discover some type issues.

### CODE REVIEW

File: modyn/metadata_database/models/selector_state_metadata.py [hunk 1]
```diff
@@ -45,7 +47,7 @@ class SelectorStateMetadata(
     __tablename__ = "selector_state_metadata"
 
     @staticmethod
-    def add_pipeline(pipeline_id: int, session: Session, engine: Engine) -> PartitionByMeta:
+    def add_pipeline(pipeline_id: int, session: Session, engine: Engine) -> PartitionByMeta | None:
```
Comment (MaxiBoether) [inline modyn/metadata_database/models/selector_state_metadata.py:50]: I think this should rather be Optional[PartitionByMeta] for consistency

File: modyn/metadata_database/models/selector_state_metadata.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
 """SelectorStateMetadata model."""
 
+from __future__ import annotations
```
Comment (MaxiBoether) [inline modyn/metadata_database/models/selector_state_metadata.py:3]: just curious why this is needed for mypy

File: modyn/metadata_processor/processor_strategies/abstract_processor_strategy.py [hunk 3]
```diff
@@ -19,7 +21,7 @@ class AbstractProcessorStrategy(ABC):
     """
 
     def __init__(self, modyn_config: dict, pipeline_id: int):
-        self.processor_strategy_type: ProcessorStrategyType = None
+        self.processor_strategy_type: ProcessorStrategyType | None = None
```
Comment (MaxiBoether) [inline modyn/metadata_processor/processor_strategies/abstract_processor_strategy.py:24]: optional

File: modyn/models/dlrm/cuda_ext/sparse_embedding.py [hunk 4]
```diff
@@ -57,7 +58,7 @@ class JointSparseEmbedding(nn.Module):
         device (torch.device): where to create the embedding. Default "cuda"
     """
 
-    def __init__(self, categorical_feature_sizes, embedding_dim, device="cuda"):
+    def __init__(self, categorical_feature_sizes: Sequence[int], embedding_dim, device="cuda"):
```
Comment (MaxiBoether) [inline modyn/models/dlrm/cuda_ext/sparse_embedding.py:61]: embedding dim and device should maybe also get typing, while we are at it

**Review by MaxiBoether — CHANGES_REQUESTED** [review]
Thank you! I had a couple of comments (mostly questions), especially on why we need `from __future__ import annotations` in many places.

File: modyn/trainer_server/internal/dataset/local_dataset_writer.py [hunk 5]
```diff
@@ -1,5 +1,8 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/trainer_server/internal/dataset/local_dataset_writer.py:1]: To allow for the `str | Path` type annotation syntax. As we are using Python 3.11+ already, this is not needed, however, mypy isn't aware of that as we are missing some configurations in the project root. I will probably look into that eventually, but I'm afraid that adding configurations could reveal many other things.

File: modyn/supervisor/internal/grpc/enums.py [hunk 6]
```diff
@@ -1,4 +1,15 @@
-from enum import StrEnum  # type: ignore # noqa: F401
+from enum import Enum
+
+
+class StrEnum(str, Enum):
```
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/supervisor/internal/grpc/enums.py:4]: Mypy had issues detecting that `enum.StrEnum` could be used as `str`. Looks like this is/was a known issue. Maybe configuring mypy to run with >=Python3.11 could solve it. But I didn't want to do that here

File: modyn/selector/internal/selector_manager.py [hunk 7]
```diff
@@ -117,7 +119,9 @@ def get_sample_keys_and_weights(
         if worker_id < 0 or worker_id >= num_workers:
             raise ValueError(f"Training {pipeline_id} has {num_workers} workers, but queried for worker {worker_id}!")
 
-        return self._selectors[pipeline_id].get_sample_keys_and_weights(trigger_id, worker_id, partition_id)
+        return self._selectors[pipeline_id].get_sample_keys_and_weights(
+            trigger_id, worker_id, partition_id
+        )  # type: ignore
```
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/selector/internal/selector_manager.py:124]: return type `(ndarray | ArrayWrapper)` != `list`. They behave the same though at runtime. Wrapping into the `list(...)` caused the same issues.

File: modyn/metadata_database/models/selector_state_metadata.py (hunk 1 above)
Comment (robinholzi) [reply modyn/metadata_database/models/selector_state_metadata.py:50]: I will add `pyupgrade` soonish. That will upgrade all old python syntax to Python 3.11+ and eliminate the inconsistencies.

File: modyn/models/dlrm/dlrm.py [hunk 8]
```diff
@@ -132,7 +132,7 @@ def embedding(self) -> Optional[torch.Tensor]:
         return self.top_model.embedding
 
     @property
-    def embedding_recorder(self) -> EmbeddingRecorder:
+    def embedding_recorder(self) -> EmbeddingRecorder:  # type: ignore
```
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/models/dlrm/dlrm.py:135]: iirc. we are returning a read-only member in a non read-only property function.

File: modyn/trainer_server/internal/dataset/local_dataset_writer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/dataset/local_dataset_writer.py:1]: haha you might be right. I did not know that mypy does not catch the A | B syntax, but it could also be that I used Union[] in many places because of this.

File: modyn/models/dlrm/dlrm.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/models/dlrm/dlrm.py:135]: hm, I guess I don't fully get the problem here, but works for now i guess

File: modyn/models/dlrm/cuda_ext/sparse_embedding.py (hunk 4 above)
Comment (robinholzi) [reply modyn/models/dlrm/cuda_ext/sparse_embedding.py:61]: Ah sure, didn't notice :) we can also force argument typing in the future :) I can do that when setting up a mypy config.

File: modyn/models/dlrm/dlrm.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/models/dlrm/dlrm.py:135]: these property functions also support write access when returning member variables. e.g. `obj. embedding_recorder = some_other_embedding_recorder`. The nested object we are returning here is read only though. It would be possible to fix that by using a private class member instead, but I thought that wouldn't be worth the effort.

File: modyn/models/resnet18/resnet18.py [hunk 9]
```diff
@@ -19,7 +19,7 @@ def __init__(self, model_configuration: dict[str, Any], device: str, amp: bool)
 
 class ResNet18Modyn(ResNet, CoresetSupportingModule):
     def __init__(self, model_configuration: dict[str, Any]) -> None:
-        super().__init__(BasicBlock, [2, 2, 2, 2], **model_configuration)
+        super().__init__(BasicBlock, [2, 2, 2, 2], **model_configuration)  # type: ignore
```
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/models/resnet18/resnet18.py:22]: same as below

File: modyn/supervisor/internal/supervisor.py [hunk 10]
```diff
@@ -357,11 +372,12 @@ def get_pipeline_status(self, pipeline_id: int) -> dict[str, Any]:
             ret["status"] = str(PipelineStatus.RUNNING)
         else:
             ret["status"] = str(PipelineStatus.EXIT)
+            p_info.process_handler.join()
 
             msg: dict[str, Any] = pipeline_stage_msg(
                 stage=PipelineStage.EXIT,
                 msg_type=MsgType.EXIT,
-                submsg=exit_submsg(p_info.process_handler.exitcode, p_info.check_for_exception()),
+                submsg=exit_submsg(p_info.process_handler.exitcode or 0, p_info.check_for_exception()),
```
[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/supervisor/internal/supervisor.py:380]: I wanted to do that initially, but then all tests except the dockerized integration tests passed.

Here are the relevant commits:
image


I tried to add a process join to ensure that the process has ended. docs say that `exitcode` is set IFF process ended. However, we end up with `None` at the end in the docker integration tests. Not sure why. As I didn't want to wait for multiple 2h long CI runs, I just went for a 0 as default. In the failure case I would expect `exitcode` to be present anyways.

[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/supervisor/internal/supervisor.py:380]: Okay, that is weird. Let's go for 0 then.

**Review by MaxiBoether — APPROVED** [review]
thanks for answering my questions :)
